### PR TITLE
✨ low-code CDK: Allow connector developers to specify the type of an added field

### PR DESCRIPTION
**Pull request by girarda**
## What
* Allow connector developers to specify the type of an added field

Example usage:
```
      - type: AddFields
        fields:
          - path:
              - transaction_id
            value: "{{ record['transaction_info']['transaction_id'] }}"
            value_type: str
```

This feature is necessary to [fix an regression introduced in source-paypal-transaction](https://airbytehq-team.slack.com/archives/C061K7D5J87/p1697769714684739) when it was migrated to low-code.
The values are implicitly casted to numbers, which is wrong because they should be treated as strings.

This should be followed up by
1. A PR fixing the paypal connector
2. An issue for exposing this field in the connector builder

## How
* Add `value_type` field to `AddedFieldDefinition` object
* The field can be any value defined in the `ValueType` enum
* The component factory converts the enum value to a type before propagating it to the `AddFields` component
* A few unrelated mypy issues were fixed along the way

## Recommended reading order
1. `airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml`
3. `airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py`
4. `airbyte-cdk/python/airbyte_cdk/sources/declarative/transformations/add_fields.py`
5. `airbyte-cdk/python/airbyte_cdk/sources/declarative/transformations/transformation.py`
6. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/stop_condition.py`
7. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/pagination_strategy.py`

## 🚨 User Impact 🚨
None. This is an additive feature.

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/resources/pull-requests-handbook)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:generateScaffolds` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

<details><summary><strong>Updating the Python CDK</strong></summary>

### Airbyter

Before merging:
- Pull Request description explains what problem it is solving
- Code change is unit tested
- Build and my-py check pass
- Smoke test the change on at least one affected connector
   - On Github: Run [this workflow](https://github.com/airbytehq/airbyte/actions/workflows/connectors_tests.yml), passing `--use-local-cdk --name=source-<connector>` as options
   - Locally: `airbyte-ci connectors --use-local-cdk --name=source-<connector> test`
- PR is reviewed and approved
      
After merging:
- [Publish the CDK](https://github.com/airbytehq/airbyte/actions/workflows/publish-cdk-command-manually.yml)
   - The CDK does not follow proper semantic versioning. Choose minor if this the change has significant user impact or is a breaking change. Choose patch otherwise.
   - Write a thoughtful changelog message so we know what was updated.
- Merge the platform PR that was auto-created for updating the Connector Builder's CDK version
   - This step is optional if the change does not affect the connector builder or declarative connectors.

</details>

### CODE REVIEW

**Comment by airbyte-oss-build-runner** [comment]
##  source-paypal-transaction test report (commit [`323d826091`](https://github.com/airbytehq/airbyte/pull/31638/commits/323d8260910c374459a361faa4e48131a23859ea)) - ✅

⏲️  Total pipeline duration: 01mn37s 

| Step                                                                      | Result   |
|:--------------------------------------------------------------------------|:---------|
| Build source-paypal-transaction docker image for platform(s) linux/x86_64 | ✅       |
| Acceptance tests                                                          | ✅       |
| Code format checks                                                        | ✅       |
| Validate metadata for source-paypal-transaction                           | ✅       |
| Connector version semver check                                            | ✅       |
| QA checks                                                                 | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/alex_typed_add_fields/1697777371/323d8260910c374459a361faa4e48131a23859ea/source-paypal-transaction/2.1.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:323d8260910c374459a361faa4e48131a23859ea)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-paypal-transaction test
```

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/stop_condition.py [hunk 1]
```diff
@@ -42,8 +42,12 @@ def next_page_token(self, response: requests.Response, last_records: List[Record
...
 
-    def reset(self):
+    def reset(self) -> None:
         self._delegate.reset()
 
     def get_page_size(self) -> Optional[int]:
         return self._delegate.get_page_size()
+
+    @property
+    def initial_token(self) -> Optional[Any]:
```
Comment (clnoll) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/stop_condition.py:52]: Why is this change needed?

**Review by clnoll — APPROVED** [review]
LGTM @girarda! Just one small question.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/stop_condition.py (hunk 1 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/stop_condition.py:52]: the method is an `abstractmethod` in the parent `PaginationStrategy` class so it must be implemented by `StopConditionPaginationStrategyDecorator`

**Review by bazarnov — APPROVED** [review]

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/stop_condition.py (hunk 1 above)
Comment (clnoll) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/paginators/strategies/stop_condition.py:52]: Okay, this is just something that was overlooked previously?

**Comment by airbyte-oss-build-runner** [comment]
##  source-paypal-transaction test report (commit [`12b94d6725`](https://github.com/airbytehq/airbyte/pull/31638/commits/12b94d67257e97a0594ce4abcca6afee07a96441)) - ❌

⏲️  Total pipeline duration: 6.49s 

| Step                                                                     | Result   |
|:-------------------------------------------------------------------------|:---------|
| Build source-paypal-transaction docker image for platform(s) linux/amd64 | ❌       |
| Code format checks                                                       | ✅       |
| Validate metadata for source-paypal-transaction                          | ✅       |
| Connector version semver check                                           | ✅       |
| QA checks                                                                | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/alex_typed_add_fields/1698073956/12b94d67257e97a0594ce4abcca6afee07a96441/source-paypal-transaction/2.1.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:12b94d67257e97a0594ce4abcca6afee07a96441)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-paypal-transaction test
```
